include/deal.II/lac/trilinos_vector.h

```diff
@@ -931,11 +931,93 @@ namespace TrilinosWrappers
 
 #endif
 
+} /* namespace TrilinosWrappers */
 
-}
+/*@}*/
 
 
-/*@}*/
+namespace internal
+{
+  template <typename> class ReinitRangeFactory;
+  template <typename> class ReinitDomainFactory;
+
+  /*
+   * A factory class internally used in linear_operator.h.
+   * Specialization for TrilinosWrappers::MPI::Vector.
+   */
+  template<>
+  class ReinitRangeFactory<TrilinosWrappers::MPI::Vector>
+  {
+  public:
+    template <typename Matrix>
+    std::function<void(TrilinosWrappers::MPI::Vector &, bool)>
+    operator()(const Matrix &matrix)
+    {
+      return [&matrix](TrilinosWrappers::MPI::Vector &v, bool fast)
+      {
+        v.reinit(matrix.range_partitioner(), fast);
+      };
```

Comment:
Ah, now I remember: The big obstacle for not having a general <code>Vector</code> template is the fact that this forces us to forward declare all possible variants of <code>reinit</code> in <code>linear_operator.h</code>.